### PR TITLE
std: add markdown library

### PR DESCRIPTION
**Pull request by kitsonk**
This PR add a markdown to HTML library to the std library based on [marked](https://github.com/markedjs/marked).

Basic usage:

```ts
import { parse } from "https://deno.land/std/convert/marked.ts";

const md = `# Hello markdown

Some basic markdown.
`;

const html = parse(md);

console.log(html); // <h1>Hello markdown</h1><p>Some basic markdown.</p>
```

### CODE REVIEW

**Comment by tunnckoCore** [comment]
I think `remark` is better option.

File: std/convert/markdown_test.ts [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import { assert, assertEquals } from "../testing/asserts.ts";
+import { test } from "../testing/mod.ts";
+import { MarkdownOptions, parse } from "./markdown.ts";
+
+const FIXTURE_DIR = "./std/convert/fixtures";
+
+const inputs = Deno.readDirSync(FIXTURE_DIR);
+const decoder = new TextDecoder();
+
+function parseOptions(str: string): [string, Partial<MarkdownOptions> | undefined] {
```
Comment (kt3k) [inline std/convert/markdown_test.ts]: I guess this function parses the front-matter of markdown. Maybe we should have another module for this later. (I think it's ok to have this here for now.)

**Comment by kevinkassimo** [comment]
Alas this actually reminded me of [denolib/marked](https://github.com/denolib/marked). Anyways this PR seems to have better type annotations and I am happy to see tests well included.

**Comment by kitsonk** [comment]
@kevinkassimo apologies, I didn't realise that was there... It was a side conversation with @ry at some point when working on the docs things saying he wanted a markdown parser as part of `std`.

File: std/convert/markdown_test.ts (hunk 1 above)
Comment (kitsonk) [reply std/convert/markdown_test.ts]: This is specific to the fixtures I ported from marked, I didn't want to change them.

**Comment by ry** [comment]
Cool - I think this is useful!

I'm not sure about the directory structure `std/convert/markdown` ... hierarchical structures are difficult because they induce extra decision making. What else is going to be in "convert" ? What about "std/encoding" ?

**Comment by nayeemrmn** [comment]
Why not just `std/markdown/mod.ts`? It stands on its own as much as other top level directories.

**Comment by kitsonk** [comment]
Away on holiday.  Encoding is fine though I would say it was stretching the term, but it is close enough and people won't get confused, as it is a similar API. Will update PR in a few days.

**Comment by nayeemrmn** [comment]
It's a big stretch. `encoding` has data languages. `markdown` is a markup language. It seems like the API mainly serves to render to HTML, not "parse" to some basic JS representation. That's a very different API from `encoding` modules.

File: std/convert/markdown.ts [hunk 2]
```diff
@@ -0,0 +1,1738 @@
...
+  return -1;
+}
+
+/**
+ * Core function which takes an input of markdown and returns HTML.
+ */
+export function parse(
+  src: string,
+  opt?: Partial<MarkdownOptions>
+): string | void {
```
Comment (nayeemrmn) [inline std/convert/markdown.ts]: This should be `renderToHtml()` or something...

There are also a lot of unneeded exports. Can we put all of this away in a `marked_es` subdirectory and export selectively under aliases? Like with [`std/encoding/yaml.ts`](https://github.com/denoland/deno/blob/f07ffe89806392698fb8052691b2826f82f15bbe/std/encoding/yaml.ts).

Comment (kitsonk) [reply std/convert/markdown.ts]: The exports match the library that it was adapted from, including the naming of this function.

Comment (nayeemrmn) [reply std/convert/markdown.ts]: Yeah, that's why I suggested giving it the vendored code treatment and abstracting over it.

File: std/encoding/mod.ts [hunk 3]
```diff
@@ -10,5 +10,6 @@ export {
   encode as encodeToHex,
   encodeToString as encodeToHexString
 } from "./hex.ts";
+export { parse as parseMarkdown, MarkdownOptions } from "./markdown.ts";
```
Comment (nayeemrmn) [inline std/encoding/mod.ts]: `parse as renderMarkdown`?

File: std/convert/markdown.ts (hunk 2 above)
Comment (nayeemrmn) [reply std/convert/markdown.ts]: I guess there's no need for this extra layer if it goes through a `mod.ts` anyway :p

File: std/encoding/mod.ts (hunk 3 above)
Comment (kitsonk) [reply std/encoding/mod.ts]: It is named as `parse` in the original library, all the other APIs are exposed as `parseXxx`, I don't see the logic personally in breaking that pattern.  Thoughts from @ry or @bartlomieju?

Comment (bartlomieju) [reply std/encoding/mod.ts]: I don't see value in having `std/encoding/mod.ts` - importing this file will import **a lot** of dependencies... That said, if we have it, let's keep it consistent to other already reexported modules.

Comment (ry) [reply std/encoding/mod.ts]: I agree

**Comment by ry** [comment]
Hey @kitsonk - sorry for taking so long on this patch (and the other!) - I think this markdown module would be better suited as a third party module -> deno.land/x/markdown .. I'm worried at the moment about expanding the scope of std beyond what Golang's std offers since we're in the middle of several big refactors of core (as always:)). What do you think?

**Comment by kitsonk** [comment]
@ry only tried to add it here because you suggested you wanted a markdown library in std, because the one you were using for `deno.land` was crap... 😝 As @kevinkassimo pointed out, a version already exists in [denolib/marked](https://github.com/denolib/marked).  I can contribute my improvements/changes there if it helps.

**Comment by ry** [comment]
@kitsonk Doh! sorry about that. Yes - I think it's preferable to merge with denolib/marked for now. This is certainly something that could be in std at some point... but at the moment I worry about having to review many fixes for this and then potentially having to add HTML sanitizers at some point.

**Comment by kitsonk** [comment]
Cool... Let's do that.  I wasn't able to find an HTML sanitizer that was easy to import/port into Deno.  I looked.  The problem is that most of them require either a browser DOM or some sort of complex virtual DOM support.

**Comment by ry** [comment]
if you put up a repo we can add it as deno.land/x/markdown

**Comment by kevinkassimo** [comment]
@ry As I mentioned in https://github.com/denoland/deno/issues/3419#issuecomment-559671442 , maybe in the future we can have a module election process to expand std (e.g. if markdown module is popular enough), each module with independent list of maintainers.

**Pull request closed without merging**